proposals/NNNN-init-accessors.md

```diff
@@ -0,0 +1,362 @@
+
+### Definite initialization of properties on `self`
+
+The semantics of an assignment inside of a type's initializer depend on whether or not all of `self` is initialized on all paths at the point of assignment. Before all of `self` is initialized, assignment to a computed property with an `init` accessor is re-written to an `init` accessor call; after `self` has been initialized, assignment to a computed property is re-written to a setter call.
+
+With this proposal, all of `self` is initialized if:
+* All stored properties are initialized on all paths, and
+* All computed properties with `init` accessors are virtually initialized on all paths.
+
+An assignment to a computed property with an `init` accessor before all of `self` is initialized will virtually initialize the computed property and initialize all of the stored properties specified in its `initializes` clause:
+
+```swift
+struct S {
+  var x1: Int
+  var x2: Int
+  var computed: Int {
+    init(newValue) initializes(x1, x2) { ... }
+  }
+
+  init() {
+    self.computed = 1 // initializes 'computed', 'x1', and 'x2'; 'self' is now fully initialized
+  }
+}
+```
+
+An assignment to a stored property before all of `self` is initialized will initialize that stored property. When all of the stored properties listed in the `initializes:` clause of a computed property with an `init` accessor have been initialized, that computed property is virtually initialized:
+
+```swift
+struct S {
+  var x1: Int
+  var x2: Int
+  var x3: Int
+  var computed: Int {
+    init(newValue) initializes(x1, x2) { ... }
+  }
+
+  init() {
+    self.x1 = 1 // initializes 'x1'; neither 'x2' or 'computed' is initialized
+    self.x2 = 1 // initializes 'x2' and 'computed'
+    self.x3 = 1 // initializes 'x3'; 'self' is now fully initialized
+  }
+}
+```
+
+An assignment to a computed property where at least one of the stored properties listed in `initializes:` is initialized, but `self` is not initialized, is an error. This prevents double-initialization of the underlying stored properties:
+
+```swift
+struct S {
+  var x: Int
+  var y: Int
+  var point: (Int, Int) {
+    init(newValue) initializes(x, y) {
+	    (self.x, self.y) = newValue
+    }
+    get { (x, y) }
+    set { (x, y) = newValue }
+  }
+
+  init(x: Int, y: Int) {
+    self.x = x // Only initializes 'x'
+    self.point = (x, y) // error: neither the `init` accessor nor the setter can be called here
+  }
+}
+```
+
+### Memberwise initializers
+
+If a struct does not declare its own initializers, it receives an implicit memberwise initializer based on the stored properties of the struct, because the storage is what needs to be initialized. Because `init` provide a preferred mechanism for initializing storage, the memberwise initializer parameter list will include any computed properties that subsume the initialization of stored properties instead of parameters for those stored properties.
```

Comment:
> Because `init` provide a preferred mechanism for initializing storage, the memberwise initializer parameter list will include any computed properties that subsume the initialization of stored properties instead of parameters for those stored properties.

I think this is a perfectly reasonable position, but I wonder if the notion that `init` accessors represent the preferred initialization strategy deserves mention as part of **Proposed solution** rather than just being a side mention in a subsection of **Detailed design**. That's a not-insignificant policy decision about the feature!

Reply:
Concretely, I think the "preferred mechanism" thing (and its relationship to the memberwise init) is the only thing in this proposal that casts doubt on what otherwise seems a perfectly reasonable use of `init` accessors: having _multiple_ computed properties which initialize the same underlying storage, e.g., `radians` and `rotations` both initializing `degrees`.

Reply:
I added more commentary in the Proposed Solution and in this section about member-wise initializers.